leaderelection/leaderelection_test.go

```diff
@@ -23,29 +23,116 @@ import (
 	"github.com/elastic/elastic-agent-libs/logp"
 	"github.com/hashicorp/go-uuid"
 	"github.com/stretchr/testify/suite"
+	"go.uber.org/goleak"
+	v1 "k8s.io/api/coordination/v1"
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	k8sFake "k8s.io/client-go/kubernetes/fake"
 	le "k8s.io/client-go/tools/leaderelection"
 	rl "k8s.io/client-go/tools/leaderelection/resourcelock"
 	"os"
 	"strings"
+	"sync"
 	"testing"
+	"time"
 )
 
 type LeaderElectionTestSuite struct {
 	suite.Suite
-	log *logp.Logger
+	log        *logp.Logger
+	wg         *sync.WaitGroup
+	manager    *Manager
+	opts       goleak.Option
+	kubeClient *k8sFake.Clientset
 }
 
 func TestLeaderElectionTestSuite(t *testing.T) {
 	s := new(LeaderElectionTestSuite)
-	s.log = logp.NewLogger("cloudbeat_leader_election_test_suite")
 	if err := logp.TestingSetup(); err != nil {
 		t.Error(err)
 	}
 
 	suite.Run(t, s)
 }
 
+func (s *LeaderElectionTestSuite) SetupTest() {
+	s.wg = &sync.WaitGroup{}
+	s.log = logp.NewLogger("cloudbeat_leader_election_test_suite")
+	s.opts = goleak.IgnoreCurrent()
+	s.kubeClient = k8sFake.NewSimpleClientset()
+	s.manager = &Manager{
+		log:        s.log,
+		leader:     nil,
+		wg:         s.wg,
+		cancelFunc: nil,
+		kubeClient: s.kubeClient,
+	}
+}
+
+func (s *LeaderElectionTestSuite) TearDownTest() {
+	// Stop is blocking until all go routines are finished,
+	// we verify there is no running leader-election managers after calling stop.
+	s.manager.Stop()
+
+	// Verify no goroutines are leaking. Safest to keep this on top of the function.
+	goleak.VerifyNone(s.T(), s.opts)
+}
+
+func (s *LeaderElectionTestSuite) TestManager_RunWaitForLeader() {
+	sTime := time.Now()
+	err := s.manager.Run(context.Background())
+	elapsed := time.Since(sTime)
```

Comment:
`Run` function is running asynchronously and gets the leader in the background.
In other words, I don't think you really measure the amount of time we waited for a leader.